<a href="https://colab.research.google.com/github/SNEHA-17/ETE456_LabProjects/blob/main/1608004_Deep_Neural_Network_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**ETE456: Implementation of Deep Neural Network on a Real World Dataset**

**Objectives:**



1.   To pre-process the data before feeding into the ML models.
2.   To apply various classification algorithms on the real world dataset.
3.   To solve Classification problems using Deep Neural Network.
4.   To develop DNN models using Tensorflow Framework. 







**Problem Statement:**
Using Deep Neural Network classifying the given dataset and developing DNN models using Tensorflow Framework.

In [ ]:
#Downloading the dataset
!wget -O diabetes.csv  https://www.dropbox.com/s/sglreynnyx40pb6/diabetes.csv?dl=0

--2021-12-16 12:58:25--  https://www.dropbox.com/s/sglreynnyx40pb6/diabetes.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.2.18, 2620:100:6022:18::a27d:4212
Connecting to www.dropbox.com (www.dropbox.com)|162.125.2.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/sglreynnyx40pb6/diabetes.csv [following]
--2021-12-16 12:58:25--  https://www.dropbox.com/s/raw/sglreynnyx40pb6/diabetes.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc6ee35792558967a3ada3803424.dl.dropboxusercontent.com/cd/0/inline/Bb_ffPBYzuy_Q5-CUk5aIrKNgKdqFi9JT4-jl8DGjZcweOyCOe65CRw62EZL-PmrnKtVU91B_JayhlGEr5Ld5Pr6vptaIpj55loWAIH3bHBjgO0wS_vgAPzDr5mbC4jxTIov5wjWpR5u9ve_9Efj66nP/file# [following]
--2021-12-16 12:58:25--  https://uc6ee35792558967a3ada3803424.dl.dropboxusercontent.com/cd/0/inline/Bb_ffPBYzuy_Q5-CUk5aIrKNgKdqFi9JT4-jl8DGjZcweOyCOe65CRw62EZL-PmrnKtVU91B_JayhlGEr5Ld

In [ ]:
import warnings
warnings.filterwarnings("ignore")

import numpy as np        
import pandas as pd     
import matplotlib.pyplot as plt       

# Machine learning libraries
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [ ]:
"""importing the dataset """

d_set = pd.read_csv('diabetes.csv')
d_set

Pregnancies  Glucose  ...  Age  Outcome
0              6      148  ...   50        1
1              1       85  ...   31        0
2              8      183  ...   32        1
3              1       89  ...   21        0
4              0      137  ...   33        1
..           ...      ...  ...  ...      ...
763           10      101  ...   63        0
764            2      122  ...   27        0
765            5      121  ...   30        0
766            1      126  ...   47        1
767            1       93  ...   23        0

[768 rows x 9 columns]

In [ ]:
d_set.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [ ]:
features = d_set[[ 'Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age']] 
target = d_set[['Outcome']]

In [ ]:
from sklearn.impute import SimpleImputer

imputer = SimpleImputer(missing_values=np.nan,strategy = "mean")  
imputer = imputer.fit(features[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age']]) 
features[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age']]= imputer.transform(features[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age']])

**Spliting the Dataset into Training Set and Test Set:**

In [ ]:
X_train,X_test,y_train,y_test=train_test_split(features,target,test_size = 0.2,random_state = 0)

(614, 8)
(154, 8)


**Logistic Regression:**

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score,f1_score,precision_score,recall_score

# Fitting Logistic Regression to the training dataset
lr = LogisticRegression()

lr.fit(X_train,y_train)

LogisticRegression()

In [ ]:
# prediction
y_pred = lr.predict(X_test)

In [ ]:
# Making confusing matrix (actual,prediction)
confusion_matrix(y_test,y_pred)

array([[98,  9],
       [18, 29]])

In [ ]:
# Making confusing matrix
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

              precision    recall  f1-score   support

          No       0.84      0.92      0.88       107
         Yes       0.76      0.62      0.68        47

    accuracy                           0.82       154
   macro avg       0.80      0.77      0.78       154
weighted avg       0.82      0.82      0.82       154



**K Nearest Neighbor:**

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

# Classifier Model
classifier = KNeighborsClassifier(n_neighbors=5, metric = 'minkowski', p=2)
classifier.fit(X_train,y_train)
# Prediction
y_pred = classifier.predict(X_test)

In [ ]:
# Making confusing matrix
print(confusion_matrix(y_test,y_pred))

# Making confusing matrix
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

[[87 20]
 [18 29]]
              precision    recall  f1-score   support

          No       0.83      0.81      0.82       107
         Yes       0.59      0.62      0.60        47

    accuracy                           0.75       154
   macro avg       0.71      0.72      0.71       154
weighted avg       0.76      0.75      0.75       154



**Support Vector Machine:**

In [ ]:
from sklearn.svm import SVC

# Classifier Model
classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(X_train,y_train)
# Prediction
y_pred = classifier.predict(X_test)

In [ ]:
# Making confusing matrix
print(confusion_matrix(y_test,y_pred))

# Making confusing matrix
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

[[97 10]
 [18 29]]
              precision    recall  f1-score   support

          No       0.84      0.91      0.87       107
         Yes       0.74      0.62      0.67        47

    accuracy                           0.82       154
   macro avg       0.79      0.76      0.77       154
weighted avg       0.81      0.82      0.81       154



**Decision Tree Classifier:**

In [ ]:
from sklearn.tree import DecisionTreeClassifier
# Classifier Model
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train,y_train)
# Prediction
y_pred = classifier.predict(X_test)

In [ ]:
# Making confusing matrix
print(confusion_matrix(y_test,y_pred))

# Making confusing matrix
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

[[79 28]
 [17 30]]
              precision    recall  f1-score   support

          No       0.82      0.74      0.78       107
         Yes       0.52      0.64      0.57        47

    accuracy                           0.71       154
   macro avg       0.67      0.69      0.67       154
weighted avg       0.73      0.71      0.72       154



**Random Forest Classiifier:**

In [ ]:
from sklearn.ensemble import RandomForestClassifier
# Classifier Model
classifier = RandomForestClassifier(n_estimators=10, criterion ='entropy', random_state = 0)
classifier.fit(X_train,y_train)
# Prediction
y_pred = classifier.predict(X_test)

In [ ]:
# Making confusing matrix
print(confusion_matrix(y_test,y_pred))

# Making confusing matrix
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

[[94 13]
 [20 27]]
              precision    recall  f1-score   support

          No       0.82      0.88      0.85       107
         Yes       0.68      0.57      0.62        47

    accuracy                           0.79       154
   macro avg       0.75      0.73      0.74       154
weighted avg       0.78      0.79      0.78       154



**Naive Bayes:**

In [ ]:
from sklearn.naive_bayes import GaussianNB

# Classifier Model
classifier = GaussianNB()
classifier.fit(X_train,y_train)
# Prediction
y_pred = classifier.predict(X_test)

In [ ]:
# Making confusing matrix
print(confusion_matrix(y_test,y_pred))

# Making confusing matrix
print(classification_report(y_test,y_pred,target_names = ['No','Yes']))

[[93 14]
 [18 29]]
              precision    recall  f1-score   support

          No       0.84      0.87      0.85       107
         Yes       0.67      0.62      0.64        47

    accuracy                           0.79       154
   macro avg       0.76      0.74      0.75       154
weighted avg       0.79      0.79      0.79       154



**Building a Multilayer Neural Network with Tensorflow Keras:**

In [ ]:
import tensorflow as tf
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

In [ ]:
#the neural network architecture
model = Sequential()
model.add(Dense(20, input_dim=8, activation='relu'))
model.add(Dense(10, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

In [ ]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 20)                180       
                                                                 
 dense_4 (Dense)             (None, 10)                210       
                                                                 
 dense_5 (Dense)             (None, 1)                 11        
                                                                 
Total params: 401
Trainable params: 401
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

**Training the Model:**

In [ ]:
#trainng the neural network on the train dataset
history = model.fit(X_train, y_train, epochs=100, validation_split=0.1)

Epoch 1/100
18/18 [==============================] - 1s 14ms/step - loss: 7.8153 - accuracy: 0.5236 - val_loss: 4.8470 - val_accuracy: 0.5323
Epoch 2/100
18/18 [==============================] - 0s 4ms/step - loss: 3.1902 - accuracy: 0.4547 - val_loss: 1.5627 - val_accuracy: 0.4032
Epoch 3/100
18/18 [==============================] - 0s 3ms/step - loss: 1.1966 - accuracy: 0.6214 - val_loss: 1.3879 - val_accuracy: 0.6290
Epoch 4/100
18/18 [==============================] - 0s 3ms/step - loss: 0.8612 - accuracy: 0.6286 - val_loss: 1.1261 - val_accuracy: 0.5484
Epoch 5/100
18/18 [==============================] - 0s 3ms/step - loss: 0.7845 - accuracy: 0.6395 - val_loss: 1.1337 - val_accuracy: 0.5806
Epoch 6/100
18/18 [==============================] - 0s 3ms/step - loss: 0.7554 - accuracy: 0.6685 - val_loss: 1.0907 - val_accuracy: 0.5323
Epoch 7/100
18/18 [==============================] - 0s 3ms/step - loss: 0.7619 - accuracy: 0.6413 - val_loss: 1.0813 - val_accuracy: 0.5806
Epoch 8/100


In [ ]:
#Prediction
y_pred = model.predict(X_test)

In [ ]:
y_pred

array([[8.4605318e-01],
       [1.7795524e-01],
       [1.8950206e-01],
       [6.6701663e-01],
       [3.1639594e-01],
       [1.5060037e-02],
       [6.2128043e-01],
       [5.4166859e-01],
       [7.4524939e-01],
       [4.7938254e-01],
       [6.2929177e-01],
       [7.1586007e-01],
       [2.3915482e-01],
       [4.4420829e-01],
       [1.2491712e-01],
       [2.8891128e-01],
       [8.5981572e-01],
       [3.1280363e-01],
       [2.9713547e-01],
       [3.9035898e-01],
       [4.2931116e-01],
       [1.9698307e-01],
       [3.9006990e-01],
       [5.4051131e-02],
       [3.7011027e-02],
       [5.7884675e-01],
       [1.5271732e-01],
       [6.1039472e-01],
       [3.2692084e-01],
       [2.9827654e-01],
       [4.2560363e-01],
       [1.8706858e-01],
       [1.6651154e-03],
       [5.8682001e-01],
       [3.0251023e-01],
       [5.9412128e-01],
       [4.1693413e-01],
       [4.6252691e-05],
       [8.2550502e-01],
       [8.6079770e-01],
       [2.3335132e-01],
       [8.523830

In [ ]:
y_pred = y_pred > 0.5
y_pred

array([[ True],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [

In [ ]:
pred = y_pred.astype('uint8')
pred

array([[1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
    

In [ ]:
#Classification Report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.79      0.83      0.81       107
           1       0.56      0.49      0.52        47

    accuracy                           0.73       154
   macro avg       0.67      0.66      0.67       154
weighted avg       0.72      0.73      0.72       154



**Conclusion:**

*   Logistic Regression: Here overall model accuracy is 82% but F1 score for classwise accuracy is 88% for No class and 68% accuracy for Yes class.
*   K Nearest Neighbor: Here overall model accuracy is 75% but F1 score for classwise accuracy is 82% for No class and 60% accuracy for Yes class.
*   Support Vector Machine: Overall model accuracy is 82% but F1 score for classwise accuracy is 87% for No class and 67% accuracy for Yes class.
*   Decision Tree Classifier: Overall model accuracy is 71% but F1 score for classwise accuracy is 78% for No class and 57% accuracy for Yes class.
*   Random Forset Classifier: Overall model accuracy = 79% .
F1 score for classwise accuracy is 85% for No class and 62% accuracy for Yes class.
*   Naive Bayes: Overall model accuracy is 79% but F1 score for classwise accuracy is 85% for No class and 64% accuracy for Yes class.
*   Neural Network: Overall model accuracy is 73% but F1 score for classwise accuracy is 81% for No class and 52% accuracy for Yes class.




